storage/replica_command.go

```diff
@@ -756,10 +756,8 @@ func (r *Replica) PushTxn(batch engine.Engine, ms *engine.MVCCStats, args proto.
 		}
 	} else {
 		// Some sanity checks for case where we don't find a transaction record.
-		if args.PusheeTxn.LastHeartbeat != nil {
-			return reply, proto.NewTransactionStatusError(&args.PusheeTxn,
-				"no txn persisted, yet intent has heartbeat")
-		} else if args.PusheeTxn.Status != proto.PENDING {
+		// If `args.PusheeTxn.LastHeartbeat != nil`, assume txn is gc'ed, see #2111.
```

Comment:
There's some value in some type of comment to discuss the case of a record being missing. Maybe above the `if ok { ...` above, with content something like

>
// There are three cases in which there is no transaction entry, in decreasing likelyhood
// * a transaction is still active; it just hasn't committed/restarted/aborted/heartbeat yet
// * the transaction resolved its intents synchronously on successful commit; in that case,
// * the transaction record is also removed. Note that in this case, the intent which prompted
//   this `PushTxn` doesn't exist any more.
// * the transaction timed out/was aborted and the intent not cleaned up, but the transaction
//  record recycled.
// We currently make no attempt at guessing which one it is, though we could (see #1939).
// Instead, a new entry is always written, and the Push assumes in all cases that the transaction is still running.
// TODO(tschottdorf): we should actually improve this when we garbage-collect aborted transactions, or we run the risk of a push recreating a GC'ed transaction as `PENDING`, which is an error if it has open intents (which is likely if someone pushes it). Likely a non-issue since the timestamp will indicate the transaction as aborted, yet should be thought out carefully.